react-redux/react-redux.d.ts

```diff
@@ -7,39 +7,68 @@
 /// <reference path="../redux/redux.d.ts" />
 
 declare module "react-redux" {
-  import { ComponentClass, Component, StatelessComponent } from 'react';
+  import { ComponentClass, Component, StatelessComponent, Props, ReactNode } from 'react';
   import { Store, Dispatch, ActionCreator } from 'redux';
 
-  export interface ComponentConstructDecorator<P> {
-    <TComponentConstruct extends (ComponentClass<P>|StatelessComponent<P>)>(component: TComponentConstruct): TComponentConstruct
+  /** Generic decorator, that receives T = original props, U = own props */
+  interface ComponentDecorator<T extends Props<any>, U extends Props<any>> {
+    (component: ComponentClass<T>): ComponentClass<U>;
+  }
+
+  /**
+   * Decorator that infers the type from the original component
+   *
+   * Can't use the above decorator because it would default the type to {}
+   */
+  export interface InferableComponentDecorator {
+    <P, TComponentConstruct extends (ComponentClass<P>|StatelessComponent<P>)>(component: TComponentConstruct): TComponentConstruct;
   }
 
   /**
    * Connects a React component to a Redux store.
+   *
+   * - Without arguments, just wraps the component, without changing the behavior / props
+   *
+   * - If 2 params are passed (3rd param, mergeProps, is skipped), default behavior
+   * is to override ownProps (as stated in the docs), so what remains is everything that's
+   * not a state or dispatch prop
+   *
+   * - When 3rd param is passed, we don't know if ownProps propagate and whether they
+   * should be valid component props, because it depends on mergeProps implementation.
+   * As such, it is the user's responsability to extend ownProps interface from state or
+   * dispatch props or both when applicable
+   *
    * @param mapStateToProps
    * @param mapDispatchToProps
    * @param mergeProps
    * @param options
-     */
-  export function connect<P>(mapStateToProps?: MapStateToProps,
-                             mapDispatchToProps?: MapDispatchToPropsFunction|MapDispatchToPropsObject,
-                             mergeProps?: MergeProps,
-                             options?: Options): ComponentConstructDecorator<P>;
+   */
+  export function connect(): InferableComponentDecorator;
+  export function connect<T extends Props<any>, U extends Props<any>, V extends Props<any>>(
+    mapStateToProps: MapStateToProps<T, V>,
+    mapDispatchToProps?: MapDispatchToPropsFunction<U, V>|MapDispatchToPropsObject
+  ): ComponentDecorator<T & U, V>;
+  export function connect<T extends Props<any>, U extends Props<any>, V extends Props<any>>(
+    mapStateToProps: MapStateToProps<T, V>,
+    mapDispatchToProps: MapDispatchToPropsFunction<U, V>|MapDispatchToPropsObject,
+    mergeProps: MergeProps<T, U, V>,
+    options?: Options
+  ): ComponentDecorator<T & U, V>;
```

Comment:
I think type parameter name like `T`, `U` and `V` looks too ambiguous and confusing since `connect` itself is not easy to understand at a first glance. Although seemingly verbose, how about renaming like this? 

- T => TStateProps
- U => TDispatchProps
- V => TOwnProps